src/compas_fab/backends/pybullet/client.py

```diff
@@ -0,0 +1,410 @@
+        return body_id
+
+    def get_base_frame(self, body_id=None):
+        body_id = self.body_id_or_default(body_id)
+        pose = pybullet.getBasePositionAndOrientation(body_id, physicsClientId=self.client_id)
+        return frame_from_pose(pose)
+
+    def get_base_name(self, body_id=None):
+        body_id = self.body_id_or_default(body_id)
+        return self.get_body_info(body_id).base_name.decode(encoding='UTF-8')
+
+    def get_link_state(self, link_id, body_id=None):
+        body_id = self.body_id_or_default(body_id)
+        return LinkState(*pybullet.getLinkState(body_id, link_id, physicsClientId=self.client_id))
+
+    def get_body_info(self, body_id=None):
+        body_id = self.body_id_or_default(body_id)
+        return BodyInfo(*pybullet.getBodyInfo(body_id, physicsClientId=self.client_id))
+
+    def get_joint_info(self, joint_id, body_id=None):
+        body_id = self.body_id_or_default(body_id)
+        return JointInfo(*pybullet.getJointInfo(body_id, joint_id, physicsClientId=self.client_id))
+
+    def get_num_joints(self, body_id=None):
+        body_id = self.body_id_or_default(body_id)
+        return pybullet.getNumJoints(body_id, physicsClientId=self.client_id)
+
+    def get_joint_ids(self, body_id=None):
+        body_id = self.body_id_or_default(body_id)
+        return list(range(self.get_num_joints(body_id)))
+
+    def get_joint_name(self, joint_id, body_id=None):
+        body_id = self.body_id_or_default(body_id)
+        return self.get_joint_info(joint_id, body_id).jointName
+
+    def get_link_name(self, link_id, body_id=None):
+        body_id = self.body_id_or_default(body_id)
+        if link_id == BASE_LINK_ID:
+            return self.get_base_name(body_id)
+        return self.get_joint_info(link_id, body_id).linkName.decode('UTF-8')
+
+    def get_link_frame(self, link_id, body_id=None):
+        body_id = self.body_id_or_default(body_id)
+        if link_id == BASE_LINK_ID:
+            return self.get_base_frame(body_id)
+        link_state = self.get_link_state(link_id, body_id)
+        pose = (link_state.worldLinkFramePosition, link_state.worldLinkFrameOrientation)
+        return frame_from_pose(pose)
+
+    # =======================================
+    def set_base_frame(self, frame, body_id=None):
+        # !!! This could lead to a discrepancy between the self._robot and what's going on in pybullet.
+        # !!! RCF would not match
+        body_id = self.body_id_or_default(body_id)
+        point, quaternion = pose_from_frame(frame)
+        pybullet.resetBasePositionAndOrientation(body_id, point, quaternion, physicsClientId=self.client_id)
+
+    def set_joint_position(self, joint_id, value, body_id=None):
+        body_id = self.body_id_or_default(body_id)
+        pybullet.resetJointState(body_id, joint_id, value, targetVelocity=0, physicsClientId=self.client_id)
+        # self.create_collision_map() ???
+
+    def set_joint_positions(self, joints, values, body_id=None):
```

Comment:
Please add docstrings to all of these ones if they are intended to be public, otherwise, prefix with `_`.